### PR TITLE
feat: check embedded interfaces

### LINKED ISSUE #106: Support checking embedded interfaces
### Preflight checklist

- [X] I agree to follow this project's [Code of Conduct](https://github.com/arch-go/arch-go/blob/main/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/arch-go/arch-go/blob/main/CONTRIBUTING.md)."
- [X] I have discussed this feature request with the community.

### Describe the background of your feature request

It is not possible to check embedded interfaces in naming rules.

Example:

```go
type Service interface {
	Creator // this will panic
	Update(s string) (int, error)
	io.Writer // this will panic
}

type Creator interface {
	Create()
}
```

Panic:
```
panic: interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident

goroutine 240 [running]:
github.com/arch-go/arch-go/internal/verifications/naming.retrieveMatchingInterfaces.func1({0x9f1bce0?, 0xc00007f300?})
        /myuser/mydir/arch-go/internal/verifications/naming/get_interfaces.go:68 +0x83b
go/ast.inspector.Visit(0xc00007f840, {0x9f1bce0?, 0xc00007f300?})
        /usr/local/go/src/go/ast/walk.go:386 +0x2b
go/ast.Walk({0x9f1a9e8?, 0xc00007f840?}, {0x9f1bce0, 0xc00007f300})
        /usr/local/go/src/go/ast/walk.go:51 +0x4c
go/ast.walkDeclList(...)
        /usr/local/go/src/go/ast/walk.go:38
go/ast.Walk({0x9f1a9e8?, 0xc00007f840?}, {0x9f1bc40, 0xc0003d5040})
        /usr/local/go/src/go/ast/walk.go:366 +0x36c5
go/ast.Inspect(...)
        /usr/local/go/src/go/ast/walk.go:397
github.com/arch-go/arch-go/internal/verifications/naming.retrieveMatchingInterfaces(0xc0003d5040, 0x9f17e68, {0xc0001a91f1, 0x7}, {0xc00077a000, 0x5de, 0x5df}, {0x0, 0x0, 0x0})
        /myuser/mydir/arch-go/internal/verifications/naming/get_interfaces.go:44 +0x125
github.com/arch-go/arch-go/internal/verifications/naming.getInterfacesMatching(0xc00016fcb0, {0xc0001a91f0?, 0x7d?})
        /myuser/mydir/arch-go/internal/verifications/naming/get_interfaces.go:31 +0x15a
github.com/arch-go/arch-go/internal/verifications/naming.checkNamingRule(0x0?, {{0xc0001a91e0?, 0x0?}, 0xc0001d2aa0?}, {{0xc0001aa640, 0x33}, {0xc00049f208, 0x71, 0x8f}})
        /myuser/mydir/arch-go/internal/verifications/naming/verifications.go:58 +0x67
github.com/arch-go/arch-go/internal/verifications/naming.CheckRule({{0xc0001aa640, 0x33}, {0xc00049f208, 0x71, 0x8f}}, {{0xc0001a91e0?, 0xc0002dff50?}, 0xc0001d2aa0?})
        /myuser/mydir/arch-go/internal/verifications/naming/verifications.go:40 +0x20f
github.com/arch-go/arch-go/internal/verifications/naming.CheckRules({{0xc0001aa640, 0x33}, {0xc00049f208, 0x71, 0x8f}}, {0xc0001a42b8, 0x3, 0x0?})
        /myuser/mydir/arch-go/internal/verifications/naming/verifications.go:19 +0xcc
github.com/arch-go/arch-go/api.(*architectureAnalysis).Execute.func4()
        /myuser/mydir/arch-go/api/architecture.go:76 +0x8d
created by github.com/arch-go/arch-go/api.(*architectureAnalysis).Execute in goroutine 1
        /myuser/mydir/arch-go/api/architecture.go:74 +0x219
```

### Describe your idea

The naming rules also could check embedded interfaces from:
- Same package
- Different module package
- External packages

### Are there any workarounds or alternatives?

No.

### Version

arch-go version 1.5.4

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by igoramorim**
## Related issue(s)
Closes #106 

## Checklist

- [x] I agree to follow this project's [Code of Conduct](../CODE_OF_CONDUCT.md).
- [x] I have read, and I am following this repository's [Contributing Guidelines](../CONTRIBUTING.md).
- [x] I have read the [Security Policy](../SECURITY.md).
- [x] I have referenced an issue describing the bug/feature request.
- [x] I have added tests that prove the correctness of my implementation.
- [ ] I have updated the documentation.

## Description

This PR implements the checking `naming rules` for embedded interfaces.

It checks the underlying type of `Method.List.Type` and retrieve the methods according to it.

### CODE REVIEW

**Comment by DGuhr** [comment]
Hey @igoramorim - thanks for your nice contribution! It looks good afaict, it just seems we have some ci issues currently (we. overhauled the ci process lately). Could you please rebase your PR on latest main, where I just [removed](https://github.com/arch-go/arch-go/commit/3b2ae28eee5d07e8fa60c7eee2f8b5e837e12e12) manually setting the ref (which I suspect is the problem here)? Thank you.

File: internal/verifications/naming/get_interfaces.go [hunk 1]
```diff
@@ -75,6 +68,103 @@ func retrieveMatchingInterfaces(
...
+		switch tp := field.Type.(type) {
+		case *ast.FuncType:
+			retrieveNormalMethods(currentInterface, field, tp, data)
+		case *ast.Ident:
+			retrieveMethodsFromEmbeddedInterface(currentInterface, tp, data, node)
+		case *ast.IndexExpr:
+			retrieveGenericsMethods(currentInterface, tp, data, node)
+		case *ast.SelectorExpr:
+			retrieveOtherPackageMethods(currentInterface, tp, node)
+		}
```
Comment (DGuhr) [inline internal/verifications/naming/get_interfaces.go:82]: could you please add a default case here, which panics with information about the used fiedl type?

Comment (igoramorim) [reply internal/verifications/naming/get_interfaces.go:85]: Just added. Thanks for the suggestion!

**Comment by igoramorim** [comment]
> ## [Codecov](https://app.codecov.io/gh/arch-go/arch-go/pull/116?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go) Report
> Attention: Patch coverage is `70.70707%` with `29 lines` in your changes missing coverage. Please review.
> 
> [Files with missing lines](https://app.codecov.io/gh/arch-go/arch-go/pull/116?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go)	Patch %	Lines
> [...fications/naming/test/pkgone/testing_interfaces.go](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&filepath=internal%2Fverifications%2Fnaming%2Ftest%2Fpkgone%2Ftesting_interfaces.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go#diff-aW50ZXJuYWwvdmVyaWZpY2F0aW9ucy9uYW1pbmcvdGVzdC9wa2dvbmUvdGVzdGluZ19pbnRlcmZhY2VzLmdv)	0.00%	[23 Missing ⚠️ ](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go)
> [internal/verifications/naming/get_interfaces.go](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&filepath=internal%2Fverifications%2Fnaming%2Fget_interfaces.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go#diff-aW50ZXJuYWwvdmVyaWZpY2F0aW9ucy9uYW1pbmcvZ2V0X2ludGVyZmFjZXMuZ28=)	91.66%	[4 Missing and 2 partials ⚠️ ](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go)
> [Files with missing lines](https://app.codecov.io/gh/arch-go/arch-go/pull/116?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go)	Coverage Δ	
> [internal/verifications/naming/utils.go](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&filepath=internal%2Fverifications%2Fnaming%2Futils.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go#diff-aW50ZXJuYWwvdmVyaWZpY2F0aW9ucy9uYW1pbmcvdXRpbHMuZ28=)	`97.77% <100.00%> (+4.92%)`	⬆️
> [internal/verifications/naming/get_interfaces.go](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&filepath=internal%2Fverifications%2Fnaming%2Fget_interfaces.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go#diff-aW50ZXJuYWwvdmVyaWZpY2F0aW9ucy9uYW1pbmcvZ2V0X2ludGVyZmFjZXMuZ28=)	`73.33% <91.66%> (+53.33%)`	⬆️
> [...fications/naming/test/pkgone/testing_interfaces.go](https://app.codecov.io/gh/arch-go/arch-go/pull/116?src=pr&el=tree&filepath=internal%2Fverifications%2Fnaming%2Ftest%2Fpkgone%2Ftesting_interfaces.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=arch-go#diff-aW50ZXJuYWwvdmVyaWZpY2F0aW9ucy9uYW1pbmcvdGVzdC9wa2dvbmUvdGVzdGluZ19pbnRlcmZhY2VzLmdv)	`0.00% <0.00%> (ø)`

@DGuhr I was wondering if I could add the `internal/verifications/naming/test/**` to the ignore section of .codecov.yml.

What do you think?

**Comment by DGuhr** [comment]
@igoramorim that makes sense. 👍  i am not completely happy with the test folders location, but that may be something for a follow up.

**Comment by igoramorim** [comment]
@DGuhr ok. I am going to add it to the .codecov.yml.

Abou the smoke tests. I understood that the failed test uses the arch-go.yml file in the root to setup the configuration.
I am not sure how to ignore the test folder I added.

If that is the case I may break the test files into smaller files so it passes the rule `maxPublicFunctionPerFile: 3`

**Comment by dadrus** [comment]
@igoramorim: I'll give you a working smoke tests config asap

**Comment by dadrus** [comment]
@igoramorim: Please excuse me for belated answer. Change the line https://github.com/igoramorim/arch-go/blob/f80d77a8b95122654a79e8e8b3a39755a51004ac/arch-go.yml#L30 from 3 to 15 for now. It will resolve the error for now. We'll need an option to express which packages should be ignored for the future. That will be addressed in another PR later.

**Comment by igoramorim** [comment]
> @igoramorim: Please excuse me for belated answer. Change the line https://github.com/igoramorim/arch-go/blob/f80d77a8b95122654a79e8e8b3a39755a51004ac/arch-go.yml#L30 from 3 to 15 for now. It will resolve the error for now. We'll need an option to express which packages should be ignored for the future. That will be addressed in another PR later.

Don't worry @dadrus!

I just changed the `maxPublicFunctionPerFile` to 15 and it looks fine. Thanks!

Now the "Security Scan" failed but this one I don't know what it is.

**Review by DGuhr — APPROVED** [review]
@igoramorim I am also sorry from my side for the lagging approval process here, I catched something bad in the meantime and had to get it out of the system first. I restarted trivy, which has some problems connecting to their db for downloading latest cve definitions lately (re-run helped). Thanks again for the contribution 👍

**Pull request merged**